### PR TITLE
[WIP]chore(gatsby-image): Migrate to Typescript

### PR DESCRIPTION
**Pull request by grgr-dkrk**
## Description
This PR migrates `packages/gatsby-image/src/index.js` and `packages/gatsby-image/src/__tests__/index.js` to TypeScript.

I also made some fixes needed for the migration.

1. I created a guard function because some functions have a mix of `Fluid | Fixed` properties.

2. In the `hasArtDirectionSupport` function, the expected type was different, so the internally used` Array.isArray` method has been moved to the appropriate place.

3. Installed the latest `@types/react` package to read the type information of the` loading` attribute of the `img` element.

## Related Issues
related to #21995

### CODE REVIEW

File: packages/gatsby-image/src/index.tsx [hunk 1]
```diff
@@ -1,7 +1,138 @@
...
+  crossOrigin?: CrossOrigin
+  loading?: Loading
+  draggable?: boolean
+  imageVariants: ImageVariants[]
+}
+
+type ImageVariants = IFixedObject | IFluidObject
+
+// If you modify these propTypes, please don't forget to update following files as well:
+// https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-image/index.d.ts
```
Comment (LekoArts) [inline packages/gatsby-image/src/index.tsx:59]: You can remove the line about index.d.ts as it then should be auto-generated

File: packages/gatsby-image/src/index.tsx [hunk 2]
```diff
@@ -1,7 +1,138 @@
...
+  className?: string | object
+  critical?: boolean
+  crossOrigin?: CrossOrigin
+  style?: CSSProperties
+  imgStyle?: CSSProperties
+  placeholderStyle?: object
+  placeholderClassName?: string
+  backgroundColor?: string | boolean
+  onLoad?: () => void
+  onError?: (event: any) => void
```
Comment (LekoArts) [inline packages/gatsby-image/src/index.tsx:80]: Can we type the `any` here?

**Comment by LekoArts** [comment]
Thanks for starting the work on this! This will definitely warrant a closer look as this is a really important component to our ecosystem.

You removed the manually typed index.d.ts file but haven't added a typegen yet. You can do the same as this: https://github.com/Eyas/gatsby/blob/ts/gatsby-link/packages/gatsby-link/package.json#L41-L45

I added some first comments but as I said this will require a closer look. Thanks for your work so far!

**Comment by grgr-dkrk** [comment]
@LekoArts Thanks! I fixed.

**Comment by grgr-dkrk** [comment]
I don't see any diffs in my local environment, but it seems like a Diff merged from another branch is appearing here.
I did Revert and Rebase of Commit, but it didn't change.
Let me rebuild the PR once. I apologize.

**Pull request closed without merging**